### PR TITLE
cdsbalancer: test cleanup part 3/N

### PR DESCRIPTION
**Pull request by easwars**
This PR cleans up aggregate cluster tests in the cds LB policy. It makes the tests verify behavior rather than internal state. Cleaning up these tests helped identify the issue described in https://github.com/grpc/grpc-go/issues/6562.

RELEASE NOTES: none

### CODE REVIEW

**Comment by easwars** [comment]
@zasweq : Please ignore the one line change in `xds/internal/balancer/clusterresolver/resource_resolver.go` for now. Once https://github.com/grpc/grpc-go/pull/6563 is merged, I will undo it.

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go [hunk 1]
```diff
@@ -18,1078 +18,650 @@ package cdsbalancer
...
-// The cluster handler receives a cluster name, handler starts a watch for that
-// cluster, xds client returns that it is a Leaf Node (EDS or LogicalDNS), not a
-// tree, so expectation that first update is written to buffer which will be
-// read by CDS LB. Then, send a new cluster update that is different, with the
-// expectation that it is also written to the update buffer to send back to CDS.
-func (s) TestSuccessCaseLeafNodeThenNewUpdate(t *testing.T) {
+// Tests the case where the cluster resource requested by the cds LB policy is
+// a leaf cluster. Subsequently the management server pushes an update to the
+// same leaf cluster. Verifies that the load balancing configuration pushed to
+// the cluster_resolver LB policy is as expected.
```
Comment (zasweq) [inline xds/internal/balancer/cdsbalancer/cluster_handler_test.go:67]: Mention that it sends two xDS updates for the same leaf cluster.

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go [hunk 2]
```diff
@@ -18,1078 +18,650 @@ package cdsbalancer
...
-	_, err := fakeClient.WaitForWatchCluster(ctx)
-	if err != nil {
-		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+// Tests the case where the cluster resource requested by the cds LB policy is
+// an aggregate cluster root pointing to two child clusters, one of type EDS and
+// the other of type LogicalDNS. Verifies that the load balancing configuration
+// pushed to the cluster_resolver LB policy is as expected. The test then
+// updates the root cluster resource requested by the cds LB policy to a leaf
+// cluster of type EDS and verifies the load balancing configuration pushed to
+// the cluster_resolver LB policy.
```
Comment (zasweq) [inline xds/internal/balancer/cdsbalancer/cluster_handler_test.go:250]: Nit: in this comment and many others, you mention "verifies the load balancing configuration". Can you please specify what exactly you're verifying in each case, or even just say verifies the load balancing configuration matches the leaf clusters or something like that.

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go [hunk 3]
```diff
@@ -18,1078 +18,650 @@ package cdsbalancer
```
Comment (zasweq) [inline xds/internal/balancer/cdsbalancer/cluster_handler_test.go:18]: Overall comment on this file: I like how you rewrote the tests to expect a child config to the cluster_resolver corresponding to the leaf clusters. However, I feel like by switching the tests over we drop a lot of the functionality wrt the cds balancer *not sending an update yet* due to the aggregate cluster graph not being filled yet (i.e. all the roots). You have a case for this from A->A not sending but drop a lot of the verifications that it doesn't send in many tests. Maybe add them by only adding management server resources for different clusters when needed, and have assertions it doesn't send before adding the management server resources for a specific cluster (instead of filling out the full tree at the beginning of the test).

Comment (zasweq) [reply xds/internal/balancer/cdsbalancer/cluster_handler_test.go:18]: A specific example of a dropped assertion: https://github.com/grpc/grpc-go/pull/6564/files#diff-6eccd90237cb4345aac6d80e2e740c5d686f42cab9799b3eaf261e392785af54L288. I think these checks are important, but luckily we can plumb it into the e2e style way as well :).

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go [hunk 4]
```diff
@@ -18,1078 +18,650 @@ package cdsbalancer
...
-	// written to the buffer to send to CDS matters, however there is no
-	// guarantee on the order it will start the watches of the children.
-	gotCluster, err = fakeClient.WaitForWatchCluster(ctx)
-	if err != nil {
-		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+// Tests the case where the cluster resource requested by the cds LB policy is
+// an aggregate cluster root pointing to two child clusters, one of type EDS and
+// the other of type LogicalDNS. Verifies that the load balancing configuration
+// pushed to the cluster_resolver LB policy is as expected. The test then
+// updates the aggregate cluster to point to a different set of child clusters
```
Comment (zasweq) [inline xds/internal/balancer/cdsbalancer/cluster_handler_test.go:162]: Rather than be vague "different set of child clusters" would prefer just a verbose technical description of what the system is. (i.e. s/set of child clusters/two child clusters both LogicalDNS).

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go [hunk 5]
```diff
@@ -18,1078 +18,650 @@ package cdsbalancer
...
-	fakeClient.InvokeWatchClusterCallback(xdsresource.ClusterUpdate{
-		ClusterType: xdsresource.ClusterTypeEDS,
-		ClusterName: edsService,
-	}, nil)
-	fakeClient.InvokeWatchClusterCallback(xdsresource.ClusterUpdate{
-		ClusterType: xdsresource.ClusterTypeLogicalDNS,
-		ClusterName: logicalDNSService,
-	}, nil)
+// Tests the case where a requested cluster resource switches between being a
+// leaf and an aggregate cluster. In each of these cases, the test verifies that
```
Comment (zasweq) [inline xds/internal/balancer/cdsbalancer/cluster_handler_test.go:319]: Same here about verbosity wrt aggregate cluster. Please specify what the aggregate cluster is pointing to.

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go [hunk 6]
```diff
@@ -18,1078 +18,650 @@ package cdsbalancer
...
-		}}); diff != "" {
-			t.Fatalf("got unexpected cluster update, diff (-got, +want): %v", diff)
-		}
-	case <-ctx.Done():
-		t.Fatal("Timed out waiting for update from update channel.")
+// Tests the scenario where an aggregate cluster exceeds the maximum depth,
+// which is 16. Verfies that the channel moves to TRANSIENT_FAILURE, and the
+// error is propagated to RPC callers. The test then modifies the graph to no
+// longer exceed maximum depth, and verifies that an RPC can be made
+// successfully.
```
Comment (zasweq) [inline xds/internal/balancer/cdsbalancer/cluster_handler_test.go:412]: s/the graph/the cluster, and mention it's right on the boundary of recursion after you update it (i.e. the maximum depth possible for the graph).

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go [hunk 7]
```diff
@@ -18,1078 +18,650 @@ package cdsbalancer
...
+	}
+	if err != nil && !strings.Contains(err.Error(), wantErr) {
+		t.Fatalf("EmptyCall() failed with err: %v, want err containing: %v", err, wantErr)
+	}
+
+	// Start a test service backend.
+	server := stubserver.StartTestService(t, nil)
+	t.Cleanup(server.Stop)
+
+	// Update the aggregate cluster resource to no longer exceed max depth.
```
Comment (zasweq) [inline xds/internal/balancer/cdsbalancer/cluster_handler_test.go:465]: Mention it's right on the boundary. And thus should be valid and root cluster eventually used for RPC.

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go [hunk 8]
```diff
@@ -18,1078 +18,650 @@ package cdsbalancer
...
-// exceeds max depth. An error should be written to the update channel.
-// Afterward, if a valid response comes in for another cluster, no update should
-// be written to the update channel, as the aggregate cluster graph is still in
-// the same error state.
-func (s) TestErrorStateWholeTree(t *testing.T) {
-	ch, fakeClient := setupTests()
-	ch.updateRootCluster("cluster0")
+// Tests the case where the aggregate cluster graph contains duplicates (A->[B,
+// C]; B->[C, D]). Verifies that the load balancing configuration pushed to the
+// cluster_resolver LB policy does not contain duplicates.
```
Comment (zasweq) [inline xds/internal/balancer/cdsbalancer/cluster_handler_test.go:546]: Also mention C higher priority than D in the config.

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go [hunk 9]
```diff
@@ -18,1078 +18,650 @@ package cdsbalancer
...
-	}, nil)
-	_, err = fakeClient.WaitForWatchCluster(ctx)
-	if err != nil {
-		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
 	}
 
-	// This shouldn't cause an update to be written to the update buffer,
-	// as cluster C has not received a cluster update yet.
```
Comment (zasweq) [inline xds/internal/balancer/cdsbalancer/cluster_handler_test.go:753]: Another example of a scenario that is lost. This is a corner case but is important because cluster C gets a callback which is unknown before specifying the same D, and tests the flip from this aggregate cluster is ready to update since the whole tree has updates (including both B and C's children) vs. it's not yet ready.

Comment (zasweq) [reply xds/internal/balancer/cdsbalancer/cluster_handler_test.go:753]: Specifically, tests this complicated algorithm/data structure of clusters seen: https://github.com/grpc/grpc-go/blob/master/xds/internal/balancer/cdsbalancer/cluster_handler.go#L215.

**Review by zasweq — CHANGES_REQUESTED** [review]
Some comments, particularly about lost assertions.

**Comment by easwars** [comment]
@zasweq : Had to force push since I had to rebase to master.

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go (hunk 4 above)
Comment (zasweq) [reply xds/internal/balancer/cdsbalancer/cluster_handler_test.go]: Oh my bad haha. Must have skimmed the last part. Sounds good, thanks.

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go [hunk 10]
```diff
@@ -18,1078 +18,704 @@ package cdsbalancer
...
+// cluster of type EDS and a different leaf cluster of type LogicalDNS and
+// verifies that the load balancing configuration pushed to the cluster_resolver
+// LB policy contains the expected discovery mechanisms corresponding to the
+// child clusters.
+func (s) TestAggregateClusterSuccess_ThenUpdateChildClusters(t *testing.T) {
+	lbCfgCh, _, _, _ := registerWrappedClusterResolverPolicy(t)
+	mgmtServer, nodeID, _, _, _, _, _ := setupWithManagementServer(t)
+
+	// Configure the management server with the aggregate cluster resource
+	// pointing to two child clusters. But don't include resource corresponding
```
Comment (zasweq) [inline xds/internal/balancer/cdsbalancer/cluster_handler_test.go:174]: Specify the child clusters, and mention you include EDS cluster. Applies to all aggregate cluster comments added in commit 3 explaining not send update yet.

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go [hunk 11]
```diff
@@ -18,1078 +18,704 @@ package cdsbalancer
...
-		t.Fatal("Timed out waiting for the cluster update to be written to the update buffer.")
+// Tests the case where the cluster resource requested by the cds LB policy is
+// an aggregate cluster root pointing to two child clusters, one of type EDS and
+// the other of type LogicalDNS. Verifies that the load balancing configuration
+// pushed to the cluster_resolver LB policy is as expected. The test then
+// updates the aggregate cluster to point to two child clusters, the same leaf
+// cluster of type EDS and a different leaf cluster of type LogicalDNS and
+// verifies that the load balancing configuration pushed to the cluster_resolver
+// LB policy contains the expected discovery mechanisms corresponding to the
+// child clusters.
```
Comment (zasweq) [inline xds/internal/balancer/cdsbalancer/cluster_handler_test.go:168]: For cases where you added *no cluster update sent* verification, please add that to top level testing comments.

**Review by zasweq — APPROVED** [review]
Thanks so much for adding back assertions. E2E tests look solid here. I remember writing the left red side two months into my tenure :).

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go (hunk 10 above)
Comment (easwars) [reply xds/internal/balancer/cdsbalancer/cluster_handler_test.go:174]: DOne.

File: xds/internal/balancer/cdsbalancer/cluster_handler_test.go (hunk 11 above)
Comment (easwars) [reply xds/internal/balancer/cdsbalancer/cluster_handler_test.go:168]: DOne.
